swarm.go

```diff
@@ -4,187 +4,169 @@ package swarm
 type Swarm struct {
-	swarm *ps.Swarm
+	// Close refcount. This allows us to fully wait for the swarm to be torn
+	// down before continuing.
+	refs sync.WaitGroup
+
 	local peer.ID
 	peers pstore.Peerstore
-	connh ConnHandler
 
-	dsync *DialSync
-	backf dialbackoff
-	dialT time.Duration // mainly for tests
+	conns struct {
+		sync.RWMutex
+		m map[peer.ID][]*Conn
+	}
 
-	dialer *conn.Dialer
+	listeners struct {
+		sync.RWMutex
+		m map[transport.Listener]struct{}
+	}
 
-	notifmu sync.RWMutex
-	notifs  map[inet.Notifiee]ps.Notifiee
+	notifs struct {
+		sync.RWMutex
+		m map[inet.Notifiee]struct{}
+	}
 
-	transports []transport.Transport
+	transports struct {
+		sync.RWMutex
+		m map[int]transport.Transport
+	}
 
-	// filters for addresses that shouldnt be dialed
-	Filters *filter.Filters
+	// new connection and stream handlers
+	connh   atomic.Value
+	streamh atomic.Value
 
-	// file descriptor rate limited
-	fdRateLimit chan struct{}
+	// dialing helpers
+	dsync   *DialSync
+	backf   DialBackoff
+	limiter *dialLimiter
+
+	// filters for addresses that shouldnt be dialed (or accepted)
+	Filters *filter.Filters
 
 	proc goprocess.Process
 	ctx  context.Context
 	bwc  metrics.Reporter
-
-	limiter *dialLimiter
-
-	protec ipnet.Protector
-}
-
-func NewSwarm(ctx context.Context, listenAddrs []ma.Multiaddr, local peer.ID,
-	peers pstore.Peerstore, bwc metrics.Reporter) (*Swarm, error) {
-	return NewSwarmWithProtector(ctx, listenAddrs, local, peers, nil, PSTransport, bwc)
 }
 
-// NewSwarm constructs a Swarm, with a Chan.
-func NewSwarmWithProtector(ctx context.Context, listenAddrs []ma.Multiaddr, local peer.ID,
-	peers pstore.Peerstore, protec ipnet.Protector, tpt pst.Transport, bwc metrics.Reporter) (*Swarm, error) {
-
-	listenAddrs, err := filterAddrs(listenAddrs)
-	if err != nil {
-		return nil, err
-	}
-
-	var wrap func(c transport.Conn) transport.Conn
-	if bwc != nil {
-		wrap = func(c transport.Conn) transport.Conn {
-			return mconn.WrapConn(bwc, c)
-		}
-	}
-
+// NewSwarm constructs a Swarm
+func NewSwarm(ctx context.Context, local peer.ID, peers pstore.Peerstore, bwc metrics.Reporter) *Swarm {
 	s := &Swarm{
-		swarm:  ps.NewSwarm(tpt),
-		local:  local,
-		peers:  peers,
-		ctx:    ctx,
-		dialT:  conn.DialTimeout,
-		notifs: make(map[inet.Notifiee]ps.Notifiee),
-		transports: []transport.Transport{
-			tcpt.NewTCPTransport(),
-			new(ws.WebsocketTransport),
+		local:   local,
+		peers:   peers,
+		bwc:     bwc,
+		Filters: filter.NewFilters(),
+
+		conns: struct {
+			sync.RWMutex
+			m map[peer.ID][]*Conn
+		}{
+			m: make(map[peer.ID][]*Conn),
+		},
+		notifs: struct {
+			sync.RWMutex
+			m map[inet.Notifiee]struct{}
+		}{
+			m: make(map[inet.Notifiee]struct{}),
+		},
+		listeners: struct {
+			sync.RWMutex
+			m map[transport.Listener]struct{}
+		}{
+			m: make(map[transport.Listener]struct{}),
+		},
+		transports: struct {
+			sync.RWMutex
+			m map[int]transport.Transport
+		}{
+			m: make(map[int]transport.Transport),
```

Comment:
I wonder if it would be better to instead do something like this:
```
s.conns.m =  make(map[peer.ID][]*Conn)
```
and so on.
This way we don't have to define those types.